docs/connect.html

```diff
@@ -543,6 +543,67 @@ <h6>ACL requirements</h6>
+    </ol>
+
+    <p>The migration script is located in <code>bin/connect-plugin-path.sh</code> and <code>bin\windows\connect-plugin-path.bat</code> of your Kafka installation. The script can migrate incompatible plugin artifacts already installed on your Connect worker's <code>plugin.path</code> by adding or modifying JAR or resource files. This is not suitable for environments using code-signing, as this may change the artifacts such that they will fail signature verification. View the built-in help with <code>--help</code>.</p>
+
+    <p>To perform a migration, first use the <code>list</code> subcommand to get an overview of the plugins available to the script. You must tell the script where to find plugins, which can be done with the repeatable <code>--worker-config</code>, <code>--plugin-path</code>, and <code>--plugin-location</code> arguments. The script will only migrate plugins present in the paths specified, so if you add plugins to your worker's classpath, then you will need to specify those plugins via one or more <code>--plugin-location</code> arguments.</p>
+
+    <p>Once you see that all incompatible plugins are included in the listing, you can proceed to dry-run the migration with <code>sync-manifests --dry-run</code>. This will perform all parts of the migration, except for writing the results of the migration to disk. Note that the <code>sync-manifests</code> command requires all specified paths to be writable, and may alter the contents of the directories. Make a backup of the specified paths, or copy them to a writable directory.</p>
+
+    <p>Ensure that you have a backup and the dry-run succeeds before removing the <code>--dry-run</code> flag and actually running the migration. If the migration fails without the <code>--dry-run</code> flag, then the partially migrated artifacts should be discarded. The migration is idempotent, so running it multiple times and on already-migrated plugins is safe. After the migration is completed, you should <a href="#connect_plugindiscovery_compatibility">verify the migration is complete</a>. These migration and verification steps may be put into a Continuous Integration process to automatically migrate plugins.</p>
+
+    <h5><a id="connect_plugindiscovery_migratesource" href="#connect_plugindiscovery_migratesource">Developers: Source Migration</a></h5>
+
+    <p>To make plugins compatible with <code>SERVICE_LOAD</code>, it is necessary to add <a href="https://docs.oracle.com/javase/8/docs/api/java/util/ServiceLoader.html">ServiceLoader</a> manifests to your source code, which should then be packaged in the release artifact. Manifests are resource files in <code>META-INF/services/</code> named after their superclass type, and contain a list of fully-qualified subclass names, one on each line. Manifests should be located in the same module as the classes they refer to, not in the corresponding test resources.</p>
```

Comment:
Is the "Manifests should be located in the same module as the classes they refer to, not in the corresponding test resources." part necessary? When do we expect users to do something like this?

Reply:
If they're using `EmbeddedConnectCluster`, adding a test manifest is sufficient to pass the tests while not actually doing the migration, so it's a false negative. Is this too detailed, do we expect developers to know the difference between main and test resources?

Reply:
Hmm... I'm leaning slightly towards this sentence being unnecessary, since we specify earlier that manifests "should then be packaged in the release artifact."

But, if you want to be explicit about this (which I'm not opposed to), I think we should frame this more clearly as a warning about a potential footgun; for example, "Be careful when testing plugin compatibility using the `hybrid_fail` mode; ensure that the service loader manifest for the plugin is included in any release artifacts produced for the plugin, instead of only included in the testing environment for the plugin" (something less wordy would be nice).

Reply:
I mentioned testing with a pre-release artifact at the end of the source migration section.

Reply:
LGTM, thanks!